scaffolds/java/gradle/wrapper/gradle-wrapper.properties

```diff
@@ -0,0 +1,5 @@
+distributionUrl=https\://services.gradle.org/distributions/gradle-7.4.2-bin.zip
```

Comment:
This one is tricky, and seems to divide the community, see [why-should-the-gradle-wrapper-be-committed-to-vcs](https://stackoverflow.com/questions/20348451/why-should-the-gradle-wrapper-be-committed-to-vcs).

**Option 1 - wrapper binary**
If we commit the wrapper, the drawback is that you have binaries in your repo.

**Option 2 - wrapper task**
To remove this drawback, we can add `wrapper {gradleVersion = '7.4.2'}` in `build.gradle`. Then the developer will have to execute `gradle wrapper` and use `./gradlew` for subsequent Gradle tasks. Which means that it requires to have a Gradle version already installed, and if it doesn't match the syntax in `build.gradle`, then he's screwed.

**Option 3 - Gradle as a prerequisite**
We can also simply add in the Readme that having Gradle 5+ installed is a prerequisite in case of no Docker, as it is for the JDK.

**Option 4 - installer script**
Because option 3 is tedious, we could provide a script that downloads and install Gradle (and, why not, Java). But isn't it the very reason of using Docker ? (and I'm not good at Shell scripting)

Personally I would go for option 1. Binaries in a repo is a problem in case it is regularly updated, which may cause issues in term of disk space use, repo cloning speed, and maybe Git perfs. In our case I don't see why we would need to upgrade to Gradle 7.4.3, then 7.5.x, then 7.6.x etc., because our `build.gradle` is concise and rarely needs changes.

If you really don't like it, then I'll go for option 3. It's a kind of way to tell to the developer "You don't want the easiness provided by Docker ? Fine, install stuffs yourself.". This repo is already opinionated: by using Shell scripts you discourage the use of Windows. So I wouldn't be afraid to discourage the avoidance of Docker as well.

A last side-remark about this: if we lead the developer to have its JDK only in Docker instead of locally, he would need to tell his IDE to find its JDK in Docker, which is not straightforward (I haven't done it yet).

Reply:
Thanks for this well-structured and insightful answer :). 

> Which means that it requires to have a Gradle version already installed, 

Indeed. This is consistent with the fact that you need an `npm` for Javascript, `cargo` for Rust and  `mix` for Elixir -- so it seems quite normal to me.

> and if it doesn't match the syntax in build.gradle, then he's screwed.

Though this *can* happen, it's quite improbable here with such a simple `build.gradle`. If it happens though, you can simply briefly rename `build.gradle` and do `gradle wrapper --gradle-version X.X`. Probably worth a line in the README though, it explain that there's an "escape hatch".

> If you really don't like it, then I'll go for option 3. It's a kind of way to tell to the developer "You don't want the easiness provided by Docker ? Fine, install stuffs yourself.".

It seems to me indeed inescapable that somebody not using docker will need to install some stuff themselves (for example: the JVM).  For me, committing the wrapper version in the `build.gradle` is more in line how dependency management is always done: don't provide the binary, but commit the version needed so that it can be downloaded.

> This repo is already opinionated: by using Shell scripts you discourage the use of Windows. 

Genuine question: in what way does providing scripts for Linux discourage the use of Windows ? 

To be clear, a PR that adds support for Windows is more than welcome. We started from our needs, not from a will to exclude anybody. I don't personally use Windows, so I'm not sure what would be needed to support Windows. If we can support both, that would be great of course !

> A last side-remark about this: if we lead the developer to have its JDK only in Docker instead of locally, he would need to tell his IDE to find its JDK in Docker, which is not straightforward (I haven't done it yet).

We included a docker wrapper for users without a proper Java IDE and that kind of stuff. We assumed that users with an IDE such as Eclipse/Intellij usually already have all they need to use gradle directly. 

We were more aiming at users that have a lightweight code editor and don't usually code in Java.

Reply:
> Though this can happen, it's quite improbable here with such a simple build.gradle.

It currently happens to all developers using Gradle 7+ and the master of this scaffold because of `testCompile`. But ok, I agree that we should now be safe for a couple of years.

I'll go for option 3 then.

> Genuine question: in what way does providing scripts for Linux discourage the use of Windows ?

AFAIK Windows doesn't natively support the execution of Shell scripts, a tool like Cygwin or Git Bash is required.